### ISSUE TITLE
Unable to Decrypt 

### ISSUE THREAD

**Issue by anajdi** [IRSgov/IDES-Data-Preparation-Dot-Net#13]
Hello, 
I received a notification from IRS however unable to decrypt. I received BAD DATA. I tried to manually decrypt using openSSL through the following commad: 
openssl rsautl -decrypt -hexdump -in 000000.00000.SL.422_Key privatekey.pam -out aeskey.txt

I Received the Following: 
Enter pass phrase for privatekey.pem:
RSA operation error
24976:error:0407109F:rsa routines:RSA_padding_check_PKCS1_type_2:pkcs decoding e
rror:.\crypto\rsa\rsa_pk1.c:273:
24976:error:04065072:rsa routines:RSA_EAY_PRIVATE_DECRYPT:padding check failed:.
\crypto\rsa\rsa_eay.c:602:

Please suggest

**Comment by MrGeneric** [comment]
Hi,
In your openSSL command you have privatekey.p**a**m. Then in the message you receive back it's looking for a pass phrase for privatekey.p**e**m. I don't know enough to know if it's supposed to be like that or not, but it stood out to me.

**Comment by anajdi** [comment]
Hi, 
Thank you for your post. After my review it is pem not pam sorry for the typo in my previous post.

**Comment by DanielCrouch** [comment]
Every time I have seen the Bad Data error, it has been a problem with the private key. For example, if I decrypt a packet and then change the private key to some random key and try again, it will give me Bad Data. And other methods, such as OpenSSL won't work for it either. So when both are bad, I suggest people double check the key being used.

**Comment by anajdi** [comment]
Thanks Daniel, 
Let's assume something went wrong during the encryption phase causing certificate mismatch, shall I encrypt and resend the payload again? Will it be refused by IRS? How shall I proceed?

**Comment by DanielCrouch** [comment]
No, it wouldn't be refused. If you take the exact same XML and package and send again, the worst you would get is an NDM error for duplicate MessageRefId. That would at least tell you that the initial file was received and we were able to process it. If it was received and we were not able to process it, you could get an NVF or some other code from the new file.

**Comment by anajdi** [comment]
Thanks Daniel, 
Below I'll elaborate more the issue I'm facing: 

I'm sure that the private key I'm using in "Sender's private key" during encryption is the same private key that I'm using in the "Receiver's private key" during decryption. My private key is a global sign certificate in .pfx format. I'm also certain that it was not revoked or updated. 
The public key used is a certificate in .cer format issued by encryption-service.services.irs.gov

Moreover, I used the same procedure with the same keys previously in 2015 but never faced decryption issues. 

Is there any changes at IRS side? How shall I proceed ?

**Comment by DanielCrouch** [comment]
Can you send the alert information from the RC001 email to lbi.fatca.ides@irs.gov? We can put in a ticket to get information from the logs which may help narrow down the error code.

**Comment by jachaos** [comment]
Hello,

Sorry to jump in unannounced. I read in the comments that you are using
private keys all the time, and a reference to a public key at one point but
I am not sure how you are using it.

In brief, to encrypt you use the private key and to decrypt the public key.

You encrypt the messages you send with your private key.

Yo decrypt the message you receive with the public key from the sender.

Could that be the issue?

**Comment by DanielCrouch** [comment]
One thing worth mentioning again is to make sure that your private key corresponds to the enrolled public key in IDES. You can use the same private key for encryption and decryption, but if it doesn't match what is in IDES, it won't work. Just as a precaution, you can export the keys from your certificate and update it in IDES and then send a new packet through. It may not be your issue, but it happens a lot. I have seen people use our sender and receiver test keys that we have in our testpacket.zip file, thinking those were the correct keys. Usually it is a case of the IDES administrator updating the certificate in IDES and then not passing that information down to the technical team that does the packaging. Again, that may not be your case at all, but it helps rule it out. But with your alert information, I should be able to at least find out if the inbound decryption worked or not.

**Comment by anajdi** [comment]
Daniel: 
if I send an email to lbi.fatca.ides@irs.gov will your receive a copy? 

Jachaos: 
Process is like following: 

**Encryption Phase: **
- Opening the .Net application posted by Github
- Uploading XML file 
- Uploading private Key (.pfx format, Global Sign 256 AES)
- Uploading public Key (.cer format same certificate enrolled within IRS website) 
- Press "Sign and Encrypt XML" 

**Uploading/Downloading Phase: **
- Sign in IRS Gateway and Unload the .Zip file 
- Download the incoming .Zip file

Decryption Phase: 
- Opening the .Net application posted by Github
- Unloading the downloaded .Zip file
- Uploading private Key (.pfx format, Global Sign 256 AES)
- Press "Decrypt Notification" 

The result is message popup "BAD DATA"

**Comment by jachaos** [comment]
Hello Anajdi

I think in the Decryption Phase should use the Public Key from the IRS
website, as you are trying to decrypt a message received from them.

2016-04-07 10:52 GMT-04:00 anajdi <notifications@github.com>:

**Comment by AndrewHartley** [comment]
Hi,

Surely the IRS would sign with their private key and encrypt with your
public key.  That way ONLY you the intended recipient can un-encrypt the
message and you would know for certain it came from the IRS.

Andrew

**Comment by DanielCrouch** [comment]
Yes, I will have access to the email. One other thing to check, make sure you decrypt into an empty folder so that the program isn't trying to grab an incorrect key. 

We use the enrolled Public Key in IDES to encrypt. So the receiver needs to use their corresponding Private Key to decrypt. Otherwise, anyone with the IRS Public Key could decrypt. This will reveal the AES key which is used to decrypt the payload file.

**Comment by anajdi** [comment]
Hello again, 
thank you for your posts and thank you for opening for us a ticket. 
After assessment and based on your comments, I have noted that the private is not correspondent to the public Key I'm using in the .NET application and the one uploaded at IRS website. They are totally different certificates, I think that's the issue. 
By the time you are reviewing the RC001 notification I'm going to do the following: 

1- Split my PFX certificate into private and public key using OpenSSL 
2- Update public key within IRS Website by uploading the new public key 
3- Go again through the encryption process and upload the same XML file uploaded before through IRS Gateway website
4- Wait IRS load and decrypt 

Is there anything wrong by doing so ?

**Comment by DanielCrouch** [comment]
No, that sounds like exactly what you need to do. I haven't seen the results of the ticket yet, but if there is a mismatch, your file would have never processed through. I would suggest following those steps and seeing what happens.

**Comment by anajdi** [comment]
I did the above steps and finally able to decrypt but received Fatca notification with NDC error.
The solution was as following: 
Download the certification offered by IDES website and use it within the .NET application. 
 
Thank you all for your comments and special thanks for Daniel :+1:
